File: react-native-nitro-player/ios/playlist/PlaylistManager.swift
//
//  PlaylistManager.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 10/12/25.
//

import Foundation
import NitroModules

/// Manages multiple playlists using AVPlayer's native playlist functionality
class PlaylistManager {
  private var playlists: [String: PlaylistModel] = [:]
  private var listeners: [(String, ([PlaylistModel], QueueOperation?) -> Void)] = []
  private var playlistListeners: [String: [(String, (PlaylistModel, QueueOperation?) -> Void)]] =
    [:]
  private var currentPlaylistId: String?
  private let queue = DispatchQueue(label: "com.margelo.nitro.nitroplayer.playlist")
  private var saveDebounceWorkItem: DispatchWorkItem?

  static let shared = PlaylistManager()

  private init() {
    loadFromFile()
  }

  /**
   * Create a new playlist
   */
  func createPlaylist(name: String, description: String? = nil, artwork: String? = nil) -> String {
    let id = UUID().uuidString
    let playlist = PlaylistModel(id: id, name: name, description: description, artwork: artwork)

    queue.sync {
      playlists[id] = playlist
    }

    scheduleSave()
    notifyPlaylistsChanged(.add)

    return id
  }

  /**
   * Delete a playlist
   */
  func deletePlaylist(playlistId: String) -> Bool {
    let removed = queue.sync {
      return playlists.removeValue(forKey: playlistId) != nil
    }

    if removed {
      if currentPlaylistId == playlistId {
        currentPlaylistId = nil
      }
      playlistListeners.removeValue(forKey: playlistId)
      scheduleSave()
      notifyPlaylistsChanged(.remove)
      return true
    }

    return false
  }

  /**
   * Update playlist metadata
   */
  func updatePlaylist(
    playlistId: String, name: String? = nil, description: String? = nil, artwork: String? = nil
  ) -> Bool {
    guard let playlist = queue.sync(execute: { playlists[playlistId] }) else {
      return false
    }

    queue.sync {
      playlists[playlistId] = PlaylistModel(
        id: playlist.id,
        name: name ?? playlist.name,
        description: description ?? playlist.description,
        artwork: artwork ?? playlist.artwork,
        tracks: playlist.tracks
      )
    }

    scheduleSave()
    notifyPlaylistChanged(playlistId, .update)
    notifyPlaylistsChanged(.update)

    return true
  }

  /**
   * Get a playlist by ID
   */
  func getPlaylist(playlistId: String) -> PlaylistModel? {
    return queue.sync {
      return playlists[playlistId]
    }
  }

  /**
   * Get all playlists
   */
  func getAllPlaylists() -> [PlaylistModel] {
    return queue.sync {
      return Array(playlists.values)
    }
  }

  /**
   * Add a track to a playlist
   */
  func addTrackToPlaylist(playlistId: String, track: TrackItem, index: Int? = nil) -> Bool {
    guard let playlist = queue.sync(execute: { playlists[playlistId] }) else {
      return false
    }

    queue.sync {
      var tracks = playlist.tracks
      if let index = index, index >= 0 && index <= tracks.count {
        tracks.insert(track, at: index)
      } else {
        tracks.append(track)
      }
      playlists[playlistId] = PlaylistModel(
        id: playlist.id,
        name: playlist.name,
        description: playlist.description,
        artwork: playlist.artwork,
        tracks: tracks
      )
    }

    scheduleSave()
    notifyPlaylistChanged(playlistId, .add)

    // Update TrackPlayerCore if this is the current playlist
    if currentPlaylistId == playlistId {
      TrackPlayerCore.shared.updatePlaylist(playlistId: playlistId)
    }

    return true
  }

  /**
   * Add multiple tracks to a playlist at once
   */
  func addTracksToPlaylist(playlistId: String, tracks: [TrackItem], index: Int? = nil) -> Bool {
    guard let playlist = queue.sync(execute: { playlists[playlistId] }) else {
      return false
    }

    queue.sync {
      var currentTracks = playlist.tracks
      if let index = index, index >= 0 && index <= currentTracks.count {
        currentTracks.insert(contentsOf: tracks, at: index)
      } else {
        currentTracks.append(contentsOf: tracks)
      }
      playlists[playlistId] = PlaylistModel(
        id: playlist.id,
        name: playlist.name,
        description: playlist.description,
        artwork: playlist.artwork,
        tracks: currentTracks
      )
    }

    scheduleSave()
    notifyPlaylistChanged(playlistId, .add)

    // Update TrackPlayerCore if this is the current playlist
    if currentPlaylistId == playlistId {
      TrackPlayerCore.shared.updatePlaylist(playlistId: playlistId)
    }

    return true
  }

  /**
   * Remove a track from a playlist
   */
  func removeTrackFromPlaylist(playlistId: String, trackId: String) -> Bool {
    guard let playlist = queue.sync(execute: { playlists[playlistId] }) else {
      return false
    }

    let removed = queue.sync {
      var tracks = playlist.tracks
      let initialCount = tracks.count
      tracks.removeAll { $0.id == trackId }
      let wasRemoved = tracks.count < initialCount

      if wasRemoved {
        playlists[playlistId] = PlaylistModel(
          id: playlist.id,
          name: playlist.name,
          description: playlist.description,
          artwork: playlist.artwork,
          tracks: tracks
        )
      }

      return wasRemoved
    }

    if removed {
      scheduleSave()
      notifyPlaylistChanged(playlistId, .remove)

      // Update TrackPlayerCore if this is the current playlist
      if currentPlaylistId == playlistId {
        TrackPlayerCore.shared.updatePlaylist(playlistId: playlistId)
      }
    }

    return removed
  }

  /**
   * Reorder a track in a playlist
   */
  func reorderTrackInPlaylist(playlistId: String, trackId: String, newIndex: Int) -> Bool {
    guard let playlist = queue.sync(execute: { playlists[playlistId] }) else {
      return false
    }

    let tracks = playlist.tracks
    guard let oldIndex = tracks.firstIndex(where: { $0.id == trackId }),
      newIndex >= 0 && newIndex < tracks.count
    else {
      return false
    }

    queue.sync {
      var reorderedTracks = tracks
      let track = reorderedTracks.remove(at: oldIndex)
      reorderedTracks.insert(track, at: newIndex)

      playlists[playlistId] = PlaylistModel(
        id: playlist.id,
        name: playlist.name,
        description: playlist.description,
        artwork: playlist.artwork,
        tracks: reorderedTracks
      )
    }

    scheduleSave()
    notifyPlaylistChanged(playlistId, .update)

    // Update TrackPlayerCore if this is the current playlist
    if currentPlaylistId == playlistId {
      TrackPlayerCore.shared.updatePlaylist(playlistId: playlistId)
    }

    return true
  }

  /**
   * Load a playlist for playback (sets it as current)
   */
  func loadPlaylist(playlistId: String) -> Bool {
    let exists = queue.sync { playlists[playlistId] != nil }
    guard exists else {
      return false
    }

    currentPlaylistId = playlistId

    // Update TrackPlayerCore
    TrackPlayerCore.shared.loadPlaylist(playlistId: playlistId)

    return true
  }

  /**
   * Get the current playlist ID
   */
  func getCurrentPlaylistId() -> String? {
    return currentPlaylistId
  }

  /**
   * Get the current playlist
   */
  func getCurrentPlaylist() -> PlaylistModel? {
    return currentPlaylistId.flatMap { id in queue.sync { playlists[id] } }
  }

  /**
   * Add a listener for playlist changes
   */
  func addPlaylistsChangeListener(listener: @escaping ([PlaylistModel], QueueOperation?) -> Void)
    -> () -> Void
  {
    let listenerId = UUID().uuidString
    queue.sync {
      listeners.append((listenerId, listener))
    }

    return {
      self.queue.sync {
        self.listeners.removeAll { $0.0 == listenerId }
      }
    }
  }

  /**
   * Add a listener for a specific playlist changes
   */
  func addPlaylistChangeListener(
    playlistId: String, listener: @escaping (PlaylistModel, QueueOperation?) -> Void
  ) -> () -> Void {
    let listenerId = UUID().uuidString
    queue.sync {
      if playlistListeners[playlistId] == nil {
        playlistListeners[playlistId] = []
      }
      playlistListeners[playlistId]?.append((listenerId, listener))
    }

    return {
      self.queue.sync {
        self.playlistListeners[playlistId]?.removeAll { $0.0 == listenerId }
      }
    }
  }

  private func notifyPlaylistsChanged(_ operation: QueueOperation?) {
    let (allPlaylists, currentListeners) = queue.sync {
      (Array(playlists.values), listeners)
    }
    DispatchQueue.main.async {
      currentListeners.forEach { $0.1(allPlaylists, operation) }
    }
  }

  private func notifyPlaylistChanged(_ playlistId: String, _ operation: QueueOperation?) {
    let result: (PlaylistModel, [(String, (PlaylistModel, QueueOperation?) -> Void)])? = queue.sync
    {
      guard let p = playlists[playlistId] else { return nil }
      return (p, playlistListeners[playlistId] ?? [])
    }

    guard let (playlist, currentListeners) = result else { return }

    DispatchQueue.main.async {
      currentListeners.forEach { $0.1(playlist, operation) }
    }
  }

  private func scheduleSave() {
    saveDebounceWorkItem?.cancel()
    let work = DispatchWorkItem { [weak self] in self?.saveToFile() }
    saveDebounceWorkItem = work
    // Use global background queue — saveToFile calls queue.sync internally,
    // which would deadlock if scheduled on queue itself.
    DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + 0.3, execute: work)
  }

  // MARK: - Persistence

  private func saveToFile() {
    do {
      let playlistsArray = queue.sync { Array(playlists.values) }
      let playlistsData = playlistsArray.map { playlist -> [String: Any] in
        return [
          "id": playlist.id,
          "name": playlist.name,
          "description": playlist.description ?? "",
          "artwork": playlist.artwork ?? "",
          "tracks": playlist.tracks.map { track -> [String: Any] in
            var trackDict: [String: Any] = [
              "id": track.id,
              "title": track.title,
              "artist": track.artist,
              "album": track.album,
              "duration": track.duration,
              "url": track.url,
            ]
            if let artwork = track.artwork, case .second(let artworkUrl) = artwork {
              trackDict["artwork"] = artworkUrl
            } else {
              trackDict["artwork"] = ""
            }
            if let extraPayload = track.extraPayload {
              trackDict["extraPayload"] = extraPayload.toDictionary()
            }
            return trackDict
          },
        ]
      }
      let wrapper: [String: Any] = [
        "playlists": playlistsData,
        "currentPlaylistId": currentPlaylistId as Any,
      ]
      let data = try JSONSerialization.data(withJSONObject: wrapper, options: [])
      try NitroPlayerStorage.write(filename: "playlists.json", data: data)
    } catch {
      NitroPlayerLogger.log("PlaylistManager", "❌ Error saving playlists - \(error)")
    }
  }

  private func loadFromFile() {
    // 1. Try new JSON file (post-migration)
    if let data = NitroPlayerStorage.read(filename: "playlists.json") {
      do {
        if let wrapper = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
          let playlistsDict = wrapper["playlists"] as? [[String: Any]] ?? []
          parsePlaylists(from: playlistsDict)
          currentPlaylistId = wrapper["currentPlaylistId"] as? String
        }
      } catch {
        NitroPlayerLogger.log("PlaylistManager", "❌ Error loading playlists - \(error)")
      }
      return
    }

    // 2. Migrate from UserDefaults (one-time, existing installs)
    if let data = UserDefaults.standard.data(forKey: "NitroPlayerPlaylists") {
      do {
        let playlistsDict = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] ?? []
        parsePlaylists(from: playlistsDict)
        currentPlaylistId = UserDefaults.standard.string(forKey: "NitroPlayerCurrentPlaylistId")
        // Remove old keys to free UserDefaults space
        UserDefaults.standard.removeObject(forKey: "NitroPlayerPlaylists")
        UserDefaults.standard.removeObject(forKey: "NitroPlayerCurrentPlaylistId")
        // Persist in new format
        saveToFile()
      } catch {
        NitroPlayerLogger.log("PlaylistManager", "❌ Error migrating playlists - \(error)")
      }
      return
    }

    // 3. Fresh install — nothing to load
  }

  private func parsePlaylists(from playlistsDict: [[String: Any]]) {
    queue.sync {
      playlists.removeAll()
      for playlistDict in playlistsDict {
        guard let id = playlistDict["id"] as? String,
          let name = playlistDict["name"] as? String
        else {
          continue
        }

        let description = playlistDict["description"] as? String
        let artwork = playlistDict["artwork"] as? String
        let tracksArray = playlistDict["tracks"] as? [[String: Any]] ?? []

        let tracks = tracksArray.compactMap { trackDict -> TrackItem? in
          guard let id = trackDict["id"] as? String,
            let title = trackDict["title"] as? String,
            let artist = trackDict["artist"] as? String,
            let album = trackDict["album"] as? String,
            let duration = trackDict["duration"] as? Double,
            let url = trackDict["url"] as? String
          else {
            return nil
          }

          let artworkString = trackDict["artwork"] as? String
          let artwork = artworkString.flatMap {
            !$0.isEmpty ? Variant_NullType_String.second($0) : nil
          }

          var extraPayload: AnyMap? = nil
          if let extraPayloadDict = trackDict["extraPayload"] as? [String: Any] {
            extraPayload = AnyMap()
            for (key, value) in extraPayloadDict {
              if let stringValue = value as? String {
                extraPayload?.setString(key: key, value: stringValue)
              } else if let doubleValue = value as? Double {
                extraPayload?.setDouble(key: key, value: doubleValue)
              } else if let intValue = value as? Int {
                extraPayload?.setDouble(key: key, value: Double(intValue))
              } else if let boolValue = value as? Bool {
                extraPayload?.setBoolean(key: key, value: boolValue)
              }
            }
          }

          return TrackItem(
            id: id,
            title: title,
            artist: artist,
            album: album,
            duration: duration,
            url: url,
            artwork: artwork,
            extraPayload: extraPayload
          )
        }

        playlists[id] = PlaylistModel(
          id: id,
          name: name,
          description: description,
          artwork: artwork,
          tracks: tracks
        )
      }
    }
  }
}

File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/playlist/PlaylistManager.kt
package com.margelo.nitro.nitroplayer.playlist

import android.content.Context
import com.margelo.nitro.core.AnyMap
import com.margelo.nitro.nitroplayer.QueueOperation
import com.margelo.nitro.nitroplayer.TrackItem
import com.margelo.nitro.nitroplayer.Variant_NullType_String
import com.margelo.nitro.nitroplayer.core.TrackPlayerCore
import com.margelo.nitro.nitroplayer.media.NitroPlayerMediaBrowserService
import com.margelo.nitro.nitroplayer.storage.NitroPlayerStorage
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Manages multiple playlists using ExoPlayer's native playlist functionality
 * Based on: https://developer.android.com/media/media3/exoplayer/playlists
 */
class PlaylistManager private constructor(
    private val context: Context,
) {
    private val playlists: MutableMap<String, Playlist> = mutableMapOf()
    private val listeners = CopyOnWriteArrayList<(List<Playlist>, QueueOperation?) -> Unit>()
    private val playlistListeners = mutableMapOf<String, CopyOnWriteArrayList<(Playlist, QueueOperation?) -> Unit>>()
    private var currentPlaylistId: String? = null

    private val saveHandler = android.os.Handler(android.os.Looper.getMainLooper())
    private val saveRunnable = Runnable { saveToFile() }

    private fun scheduleSave() {
        saveHandler.removeCallbacks(saveRunnable)
        saveHandler.postDelayed(saveRunnable, 300)
    }

    companion object {
        @Volatile
        @Suppress("ktlint:standard:property-naming")
        private var INSTANCE: PlaylistManager? = null

        // Legacy SharedPreferences name (migration only)
        private const val LEGACY_PREFS_NAME = "NitroPlayerPlaylists"

        @JvmStatic
        fun getInstance(context: Context): PlaylistManager =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: PlaylistManager(context.applicationContext).also { INSTANCE = it }
            }
    }

    init {
        // Don't load from file on init - only load when Android Auto needs it
    }

    /**
     * Create a new playlist
     */
    fun createPlaylist(
        name: String,
        description: String? = null,
        artwork: String? = null,
    ): String {
        val id = UUID.randomUUID().toString()
        val playlist = Playlist(id, name, description, artwork)

        synchronized(playlists) {
            playlists[id] = playlist
        }

        // Only cache for Android Auto if connected
        if (NitroPlayerMediaBrowserService.isAndroidAutoConnected) {
            scheduleSave()
        }
        notifyPlaylistsChanged(QueueOperation.ADD)
        NitroPlayerMediaBrowserService.getInstance()?.onPlaylistsUpdated()

        return id
    }

    /**
     * Delete a playlist
     */
    fun deletePlaylist(playlistId: String): Boolean {
        val removed =
            synchronized(playlists) {
                playlists.remove(playlistId)
            }

        if (removed != null) {
            if (currentPlaylistId == playlistId) {
                currentPlaylistId = null
            }
            playlistListeners.remove(playlistId)
            // Only cache for Android Auto if connected
            if (NitroPlayerMediaBrowserService.isAndroidAutoConnected) {
                scheduleSave()
            }
            notifyPlaylistsChanged(QueueOperation.REMOVE)
            NitroPlayerMediaBrowserService.getInstance()?.onPlaylistsUpdated()
            return true
        }

        return false
    }

    /**
     * Update playlist metadata
     */
    fun updatePlaylist(
        playlistId: String,
        name: String? = null,
        description: String? = null,
        artwork: String? = null,
    ): Boolean {
        synchronized(playlists) {
            val playlist = playlists[playlistId] ?: return false
            playlists[playlistId] =
                playlist.copy(
                    name = name ?: playlist.name,
                    description = description ?: playlist.description,
                    artwork = artwork ?: playlist.artwork,
                )
        }

        // Only cache for Android Auto if connected
        if (NitroPlayerMediaBrowserService.isAndroidAutoConnected) {
            scheduleSave()
        }
        notifyPlaylistChanged(playlistId, QueueOperation.UPDATE)
        notifyPlaylistsChanged(QueueOperation.UPDATE)
        NitroPlayerMediaBrowserService.getInstance()?.onPlaylistsUpdated()

        return true
    }

    /**
     * Get a playlist by ID
     */
    fun getPlaylist(playlistId: String): Playlist? =
        synchronized(playlists) {
            playlists[playlistId]
        }

    /**
     * Get all playlists
     */
    fun getAllPlaylists(): List<Playlist> =
        synchronized(playlists) {
            playlists.values.toList()
        }

    /**
     * Add a track to a playlist
     */
    fun addTrackToPlaylist(
        playlistId: String,
        track: TrackItem,
        index: Int? = null,
    ): Boolean {
        synchronized(playlists) {
            val playlist = playlists[playlistId] ?: return false
            val tracks = playlist.tracks.toMutableList()
            if (index != null && index >= 0 && index <= tracks.size) {
                tracks.add(index, track)
            } else {
                tracks.add(track)
            }
            playlists[playlistId] = playlist.copy(tracks = tracks)
        }

        // Only cache for Android Auto if connected
        if (NitroPlayerMediaBrowserService.isAndroidAutoConnected) {
            scheduleSave()
        }
        notifyPlaylistChanged(playlistId, QueueOperation.ADD)
        NitroPlayerMediaBrowserService.getInstance()?.onPlaylistUpdated(playlistId)

        // Update ExoPlayer if this is the current playlist
        if (currentPlaylistId == playlistId) {
            TrackPlayerCore.getInstance(context)?.updatePlaylist(playlistId)
        }

        return true
    }

    /**
     * Add multiple tracks to a playlist at once
     */
    fun addTracksToPlaylist(
        playlistId: String,
        tracks: List<TrackItem>,
        index: Int? = null,
    ): Boolean {
        synchronized(playlists) {
            val playlist = playlists[playlistId] ?: return false
            val currentTracks = playlist.tracks.toMutableList()
            if (index != null && index >= 0 && index <= currentTracks.size) {
                currentTracks.addAll(index, tracks)
            } else {
                currentTracks.addAll(tracks)
            }
            playlists[playlistId] = playlist.copy(tracks = currentTracks)
        }

        // Only cache for Android Auto if connected
        if (NitroPlayerMediaBrowserService.isAndroidAutoConnected) {
            scheduleSave()
        }
        notifyPlaylistChanged(playlistId, QueueOperation.ADD)
        NitroPlayerMediaBrowserService.getInstance()?.onPlaylistUpdated(playlistId)

        // Update ExoPlayer if this is the current playlist
        if (currentPlaylistId == playlistId) {
            TrackPlayerCore.getInstance(context)?.updatePlaylist(playlistId)
        }

        return true
    }

    /**
     * Remove a track from a playlist
     */
    fun removeTrackFromPlaylist(
        playlistId: String,
        trackId: String,
    ): Boolean {
        val removed =
            synchronized(playlists) {
                val playlist = playlists[playlistId] ?: return false
                val tracks = playlist.tracks.toMutableList()
                val removed = tracks.removeAll { it.id == trackId }
                if (removed) {
                    playlists[playlistId] = playlist.copy(tracks = tracks)
                }
                removed
            }

        if (removed) {
            scheduleSave()
            notifyPlaylistChanged(playlistId, QueueOperation.REMOVE)
            NitroPlayerMediaBrowserService.getInstance()?.onPlaylistUpdated(playlistId)

            // Update ExoPlayer if this is the current playlist
            if (currentPlaylistId == playlistId) {
                TrackPlayerCore.getInstance(context)?.updatePlaylist(playlistId)
            }
        }

        return removed
    }

    /**
     * Reorder a track in a playlist
     */
    fun reorderTrackInPlaylist(
        playlistId: String,
        trackId: String,
        newIndex: Int,
    ): Boolean {
        synchronized(playlists) {
            val playlist = playlists[playlistId] ?: return false
            val tracks = playlist.tracks.toMutableList()
            val oldIndex = tracks.indexOfFirst { it.id == trackId }
            if (oldIndex < 0 || newIndex < 0 || newIndex >= tracks.size) {
                return false
            }
            val track = tracks.removeAt(oldIndex)
            tracks.add(newIndex, track)
            playlists[playlistId] = playlist.copy(tracks = tracks)
        }

        scheduleSave()
        notifyPlaylistChanged(playlistId, QueueOperation.UPDATE)
        NitroPlayerMediaBrowserService.getInstance()?.onPlaylistUpdated(playlistId)

        // Update ExoPlayer if this is the current playlist
        if (currentPlaylistId == playlistId) {
            TrackPlayerCore.getInstance(context)?.updatePlaylist(playlistId)
        }

        return true
    }

    /**
     * Update entire track objects across all playlists
     * Matches by track.id and replaces the entire track object
     * @param tracks List of full TrackItem objects to update
     * @return Map of playlistId -> count of tracks updated
     */
    fun updateTracks(tracks: List<TrackItem>): Map<String, Int> {
        val tracksMap = tracks.associateBy { it.id }
        val affectedPlaylists = mutableMapOf<String, Int>()

        synchronized(playlists) {
            playlists.forEach { (playlistId, playlist) ->
                var updateCount = 0
                val newTracks =
                    playlist.tracks
                        .map { track ->
                            tracksMap[track.id]?.also { updateCount++ } ?: track
                        }.toMutableList()

                if (updateCount > 0) {
                    affectedPlaylists[playlistId] = updateCount
                    playlists[playlistId] = playlist.copy(tracks = newTracks)
                }
            }
        }

        if (affectedPlaylists.isNotEmpty()) {
            scheduleSave()
            affectedPlaylists.keys.forEach { playlistId ->
                notifyPlaylistChanged(playlistId, QueueOperation.UPDATE)
            }
            notifyPlaylistsChanged(QueueOperation.UPDATE)
        }

        return affectedPlaylists
    }

    /**
     * Get tracks by IDs from all playlists
     * @param trackIds List of track IDs to fetch
     * @return List of matching TrackItem objects
     */
    fun getTracksById(trackIds: List<String>): List<TrackItem> {
        val trackIdSet = trackIds.toSet()
        val foundTracks = mutableMapOf<String, TrackItem>()

        synchronized(playlists) {
            playlists.values.forEach { playlist ->
                playlist.tracks.forEach { track ->
                    if (trackIdSet.contains(track.id) && !foundTracks.containsKey(track.id)) {
                        foundTracks[track.id] = track
                    }
                }
            }
        }

        // Return in same order as requested
        return trackIds.mapNotNull { foundTracks[it] }
    }

    /**
     * Load a playlist for playback (sets it as current)
     */
    fun loadPlaylist(playlistId: String): Boolean {
        val playlist =
            synchronized(playlists) {
                playlists[playlistId]
            } ?: return false

        currentPlaylistId = playlistId
        TrackPlayerCore.getInstance(context)?.loadPlaylist(playlistId)

        return true
    }

    /**
     * Get the current playlist ID
     */
    fun getCurrentPlaylistId(): String? = currentPlaylistId

    /**
     * Get the current playlist
     */
    fun getCurrentPlaylist(): Playlist? = currentPlaylistId?.let { synchronized(playlists) { playlists[it] } }

    /**
     * Add a listener for playlist changes
     */
    fun addPlaylistsChangeListener(listener: (List<Playlist>, QueueOperation?) -> Unit): () -> Unit {
        listeners.add(listener)
        return { listeners.remove(listener) }
    }

    /**
     * Add a listener for a specific playlist changes
     */
    fun addPlaylistChangeListener(
        playlistId: String,
        listener: (Playlist, QueueOperation?) -> Unit,
    ): () -> Unit {
        val playlistListeners = playlistListeners.getOrPut(playlistId) { CopyOnWriteArrayList() }
        playlistListeners.add(listener)
        return { playlistListeners.remove(listener) }
    }

    private fun notifyPlaylistsChanged(operation: QueueOperation?) {
        val allPlaylists =
            synchronized(playlists) {
                playlists.values.toList()
            }
        listeners.forEach { it(allPlaylists, operation) }
    }

    private fun notifyPlaylistChanged(
        playlistId: String,
        operation: QueueOperation?,
    ) {
        val playlist =
            synchronized(playlists) {
                playlists[playlistId]
            } ?: return

        playlistListeners[playlistId]?.forEach { it(playlist, operation) }
    }

    // MARK: - Persistence

    private fun saveToFile() {
        try {
            val jsonArray = JSONArray()
            synchronized(playlists) {
                playlists.values.forEach { playlist ->
                    val jsonObject =
                        JSONObject().apply {
                            put("id", playlist.id)
                            put("name", playlist.name)
                            put("description", playlist.description ?: "")
                            put("artwork", playlist.artwork ?: "")
                            val tracksArray = JSONArray()
                            playlist.tracks.forEach { track ->
                                tracksArray.put(
                                    JSONObject().apply {
                                        put("id", track.id)
                                        put("title", track.title)
                                        put("artist", track.artist)
                                        put("album", track.album)
                                        put("duration", track.duration)
                                        put("url", track.url)
                                        track.artwork?.let { put("artwork", it) }
                                        track.extraPayload?.let { payload ->
                                            val extraPayloadMap = payload.toHashMap()
                                            val extraPayloadJson = JSONObject(extraPayloadMap)
                                            put("extraPayload", extraPayloadJson)
                                        }
                                    },
                                )
                            }
                            put("tracks", tracksArray)
                        }
                    jsonArray.put(jsonObject)
                }
            }

            val wrapper =
                JSONObject().apply {
                    put("playlists", jsonArray)
                    put("currentPlaylistId", currentPlaylistId)
                }
            NitroPlayerStorage.write(context, "playlists.json", wrapper.toString())
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun loadPlaylistsFromFile() {
        // 1. Try new JSON file (post-migration)
        val json = NitroPlayerStorage.read(context, "playlists.json")
        if (json != null) {
            try {
                val wrapper = JSONObject(json)
                val jsonArray = wrapper.optJSONArray("playlists") ?: JSONArray()
                parseAndLoadPlaylists(jsonArray)
                currentPlaylistId =
                    if (wrapper.isNull("currentPlaylistId")) {
                        null
                    } else {
                        wrapper.optString("currentPlaylistId", null.toString()).takeIf { it != "null" }
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return
        }

        // 2. Migrate from SharedPreferences (one-time, existing installs)
        val prefs = context.getSharedPreferences(LEGACY_PREFS_NAME, Context.MODE_PRIVATE)
        val legacyJson = prefs.getString("playlists", null)
        if (legacyJson != null) {
            try {
                val jsonArray = JSONArray(legacyJson)
                parseAndLoadPlaylists(jsonArray)
                currentPlaylistId = prefs.getString("currentPlaylistId", null)
                // Remove old SharedPreferences data to free space
                prefs
                    .edit()
                    .remove("playlists")
                    .remove("currentPlaylistId")
                    .apply()
                // Persist in new format
                saveToFile()
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return
        }

        // 3. Fresh install — nothing to load
    }

    private fun parseAndLoadPlaylists(jsonArray: JSONArray) {
        synchronized(playlists) {
            playlists.clear()
            for (i in 0 until jsonArray.length()) {
                val jsonObject = jsonArray.getJSONObject(i)
                val tracks = mutableListOf<TrackItem>()
                val tracksArray = jsonObject.getJSONArray("tracks")
                for (j in 0 until tracksArray.length()) {
                    val trackObj = tracksArray.getJSONObject(j)
                    val artworkStr = trackObj.optString("artwork")
                    val artwork: Variant_NullType_String? =
                        if (!artworkStr.isNullOrEmpty()) {
                            Variant_NullType_String.create(artworkStr)
                        } else {
                            null
                        }
                    val extraPayload: AnyMap? =
                        if (trackObj.has("extraPayload")) {
                            val extraPayloadJson = trackObj.getJSONObject("extraPayload")
                            val map = AnyMap()
                            val keyIterator = extraPayloadJson.keys()
                            while (keyIterator.hasNext()) {
                                val key = keyIterator.next()
                                when (val value = extraPayloadJson.get(key)) {
                                    is String -> map.setString(key, value)
                                    is Number -> map.setDouble(key, value.toDouble())
                                    is Boolean -> map.setBoolean(key, value)
                                }
                            }
                            map
                        } else {
                            null
                        }
                    tracks.add(
                        TrackItem(
                            id = trackObj.getString("id"),
                            title = trackObj.getString("title"),
                            artist = trackObj.getString("artist"),
                            album = trackObj.getString("album"),
                            duration = trackObj.getDouble("duration"),
                            url = trackObj.getString("url"),
                            artwork = artwork,
                            extraPayload = extraPayload,
                        ),
                    )
                }
                val descriptionStr = jsonObject.optString("description")
                val artworkStr = jsonObject.optString("artwork")
                val playlist =
                    Playlist(
                        id = jsonObject.getString("id"),
                        name = jsonObject.getString("name"),
                        description = if (!descriptionStr.isNullOrEmpty()) descriptionStr else null,
                        artwork = if (!artworkStr.isNullOrEmpty()) artworkStr else null,
                        tracks = tracks,
                    )
                playlists[playlist.id] = playlist
            }
        }
    }
}
